gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/webauthn/WebAuthnResponseHandler.java

```diff
@@ -0,0 +1,209 @@
+                                });
+                            });
+                        } else {
+                            logger.error("Unexpected exception", authenticate.cause());
+                            ctx.fail(authenticate.cause());
+                        }
+                    });
+        } catch (IllegalArgumentException e) {
+            logger.error("Unexpected exception", e);
+            ctx.fail(400);
+        } catch (RuntimeException e) {
+            logger.error("Unexpected exception", e);
+            ctx.fail(e);
+        }
+    }
+
+    private AuthenticationContext createAuthenticationContext(RoutingContext context) {
+        HttpServerRequest httpServerRequest = context.request();
+        SimpleAuthenticationContext authenticationContext = new SimpleAuthenticationContext(new VertxHttpServerRequest(httpServerRequest.getDelegate()));
+        authenticationContext.set(Claims.ip_address, RequestUtils.remoteAddress(httpServerRequest));
+        authenticationContext.set(Claims.user_agent, RequestUtils.userAgent(httpServerRequest));
+        authenticationContext.set(Claims.domain, domain.getId());
+        return authenticationContext;
+    }
+
+    private void authenticateUser(Client client,
+                                  AuthenticationContext authenticationContext,
+                                  String username,
+                                  String credentialId,
+                                  Handler<AsyncResult<User>> handler) {
+        credentialService.findByCredentialId(ReferenceType.DOMAIN, domain.getId(), credentialId)
+                .firstElement()
+                .switchIfEmpty(Maybe.error(new CredentialNotFoundException(credentialId)))
+                .flatMapSingle(credential -> userAuthenticationManager.connectPreAuthenticatedUser(client, credential.getUserId(), new EndUserAuthentication(username, null, authenticationContext)))
+                .subscribe(
+                        user -> handler.handle(Future.succeededFuture(new io.gravitee.am.gateway.handler.common.vertx.web.auth.user.User(user))),
+                        error -> handler.handle(Future.failedFuture(error))
+                );
+    }
+
+    private void updateCredential(AuthenticationContext authenticationContext, String credentialId, String userId, Handler<AsyncResult<Void>> handler) {
+        Credential credential = new Credential();
+        credential.setUserId(userId);
+        credential.setUserAgent(String.valueOf(authenticationContext.get(Claims.user_agent)));
+        credential.setIpAddress(String.valueOf(authenticationContext.get(Claims.ip_address)));
+        credentialService.update(ReferenceType.DOMAIN, domain.getId(), credentialId, credential)
+                .subscribe(
+                        () -> handler.handle(Future.succeededFuture()),
+                        error -> handler.handle(Future.failedFuture(error))
+                );
+    }
```

Comment:
Duplicate method that can be moved into the WebAuthnHandler

Reply:
FP